staging/kos/pkg/controllers/connectionagent/connectionagent.go

```diff
@@ -0,0 +1,1059 @@
+		// Fix this: if a selector appears both in options.FieldSelector and customFieldSelector it must
+		// appear in the resulting selector only once.
+		optionsFieldSelector := options.FieldSelector
+		allSelectors := make([]string, 0, 2)
+		if strings.Trim(optionsFieldSelector, " ") != "" {
+			allSelectors = append(allSelectors, optionsFieldSelector)
+		}
+		allSelectors = append(allSelectors, customFieldSelector)
+		options.FieldSelector = strings.Join(allSelectors, ",")
+	}
+}
+
+func v1a1AttsCustomInformerAndLister(kcs *kosclientset.Clientset,
+	resyncPeriod time.Duration,
+	tweakListOptionsFunc kosinternalifcs.TweakListOptionsFunc) (k8scache.SharedIndexInformer, koslisterv1a1.NetworkAttachmentLister) {
+
+	attv1a1Informer := createAttsv1a1Informer(kcs,
+		resyncPeriod,
+		k8smetav1.NamespaceAll,
+		tweakListOptionsFunc)
+	return attv1a1Informer.Informer(), attv1a1Informer.Lister()
+}
+
+func v1a1AttsCustomNamespaceInformerAndLister(kcs *kosclientset.Clientset,
+	resyncPeriod time.Duration,
+	namespace string,
+	tweakListOptionsFunc kosinternalifcs.TweakListOptionsFunc) (k8scache.SharedIndexInformer, koslisterv1a1.NetworkAttachmentNamespaceLister) {
+
+	attv1a1Informer := createAttsv1a1Informer(kcs,
+		resyncPeriod,
+		namespace,
+		tweakListOptionsFunc)
+	return attv1a1Informer.Informer(), attv1a1Informer.Lister().NetworkAttachments(namespace)
+}
+
+func createAttsv1a1Informer(kcs *kosclientset.Clientset,
+	resyncPeriod time.Duration,
+	namespace string,
+	tweakListOptionsFunc kosinternalifcs.TweakListOptionsFunc) kosinformersv1a1.NetworkAttachmentInformer {
+
+	localAttsInformerFactory := kosinformers.NewFilteredSharedInformerFactory(kcs,
+		resyncPeriod,
+		namespace,
+		tweakListOptionsFunc)
+	netv1a1Ifc := localAttsInformerFactory.Network().V1alpha1()
+	return netv1a1Ifc.NetworkAttachments()
+}
+
+// attachmentMACAddr is an Index function that computes the MAC address of a NetworkAttachment
+func attachmentMACAddr(obj interface{}) (subnets []string, err error) {
+	att := obj.(*netv1a1.NetworkAttachment)
+	return []string{generateMACAddr(att.Spec.VNI, gonet.ParseIP(att.Status.IPv4)).String()}, nil
+}
+
+// TODO factor out attNSN in pkg controller/utils, as it is used both by
+// the connection agent and the IPAM controller.
+func attNSN(netAtt *netv1a1.NetworkAttachment) k8stypes.NamespacedName {
+	return k8stypes.NamespacedName{Namespace: netAtt.Namespace,
+		Name: netAtt.Name}
+}
+
+func generateIfcName(macAddr gonet.HardwareAddr) string {
+	return "kos" + macAddr.String()
+}
+
+func generateMACAddr(vni uint32, guestIPv4 gonet.IP) gonet.HardwareAddr {
+	guestIPBytes := guestIPv4.To4()
+	mac := make([]byte, 6, 6)
+	mac[5] = byte(vni >> 0)
+	mac[4] = byte(vni >> 8)
+	mac[3] = byte(vni>>16) | (guestIPBytes[3] << 5)
+	mac[2] = (guestIPBytes[3] >> 3) | (guestIPBytes[2] << 5)
+	mac[1] = (guestIPBytes[2] >> 3) | (guestIPBytes[1] << 5)
+	mac[0] = (guestIPBytes[1] & 0xF8) | (guestIPBytes[0] & 0x04) | 2
```

Comment:
This encoding produces collisions.  For example, between 10.168.1.1 and 192.168.1.1.
For the 25th bit of address you want something like `( ( (guestIPBytes[0]&8) >> 3) + (guestIPBytes[0]&4) >> 2) ) & 1`.
Also, how about doing less shifting and putting whole bytes directly in `mac[2]` and `mac[1]`?

Reply:
Bug fixed in a597f1e.